aten/src/ATen/native/cuda/AdaptiveAveragePooling.cu

```diff
@@ -274,47 +561,120 @@ namespace {
     checkAllSameGPU("cudnn_adaptive_avg_pooling2d_out",
                     {grad_input_arg, grad_output_arg, input_arg});
 
-    bool atomic = true; // suboptimal, but without atomic it doesn't pass the tests
-
-    Tensor gradOutput = gradOutput_.contiguous();
+    if (input.suggest_memory_format() == at::MemoryFormat::ChannelsLast) {
+      // special case for tensor memory format in channels_last
+      TORCH_CHECK(input.ndimension() == 4,
+        "non-empty 4D (batch mode) tensor expected for input with channels_last layout");
 
-    int64_t sizeD  = input.size(-3);
-    int64_t isizeH = input.size(-2);
-    int64_t isizeW = input.size(-1);
+      int sizeB = input.size(0);
+      int sizeC = input.size(1);
+      int isizeH = input.size(2);
+      int isizeW = input.size(3);
 
-    int64_t osizeH = gradOutput.size(-2);
-    int64_t osizeW = gradOutput.size(-1);
+      Tensor gradOutput = gradOutput_;
 
-    int64_t grid_x = sizeD;
-    if (input.ndimension() == 4) grid_x *= input.size(-4);
+      int64_t ostrideB = gradOutput.stride(0);
+      int64_t ostrideC = gradOutput.stride(1);
+      int64_t ostrideH = gradOutput.stride(2);
+      int64_t ostrideW = gradOutput.stride(3);
 
-      //bool atomic = (isizeW%osizeW != 0) || (isizeH%osizeH != 0);
-    AT_DISPATCH_FLOATING_TYPES_AND_HALF(
-        input.scalar_type(), "adaptive_avg_pool2d_backward_cuda", [&] {
-          scalar_t *gradOutput_data = gradOutput.data_ptr<scalar_t>();
-          scalar_t *gradInput_data = gradInput.data_ptr<scalar_t>();
+      int osizeH = gradOutput.size(-2);
+      int osizeW = gradOutput.size(-1);
 
-          // cuda blocks & threads:
-          int blocksH = std::max((int)(16L / sizeD), 1);
-          dim3 blocks(grid_x, blocksH);
-          dim3 threads(32, 8);
+      // preserve channels_last stride on input tensor;
+      if (!gradInput.is_contiguous(at::MemoryFormat::ChannelsLast)) {
```

Comment:
Then there will be a silent error. (illegal memory access that can't be caught).

In current implementation, `gradInput` here will always be packed and if that error is to occur it will be at a level higher: https://github.com/pytorch/pytorch/pull/24396/files/1f9f43525b9d71b18184965000acce37e0bab2cd#diff-7b64cf245203cfc8f04802bd74787113R708
Do we want to put an assert there?

Reply:
Assert or inplace update would be nice.

Reply:
Oops, I stand corrected.

In case of non-packed memory space, `resize_` would just change memory stride as needed (or allocate new storage if necessary, this would also affect the base memory dense tensor):
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Resize.cuh

This does not make much sense to me, but at least we are not getting `illegal memory access` and we are keeping the default behavior.
We can update the behavior if we think it's no good, but I would prefer to do that in a separate PR as it's not related to the memory format support that we are adding here.